### PR TITLE
Build config docs

### PR DESCRIPTION
**Pull request by jakemac53**
I think eventually it would be cool to auto-gen these docs using json serializable, and this was initially an effort to see what that format might eventually look like - but I think its worth submitting.

The only real downside is we might forget to update it if we change the format but its easy to fix.

### CODE REVIEW

File: build_config/README.md [hunk 1]
```diff
@@ -3,6 +3,10 @@
 Customizing the build behavior of a package is done  by creating a `build.yaml`
 file, which describes your configuration.
 
+The full format is described in the
+[docs/build_yaml_format.md](../docs/build_yaml_format.md) file, while this
```
Comment (natebosch) [inline build_config/README.md]: [nit] this link won't work from the README on the pub site. Maybe add a full URL?

File: docs/build_yaml_format.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+include | List<String> | `**`
+exclude | List<String> | none
+
+## TargetBuilderConfig
+
+key | value | default
+--- | --- | ---
+enabled | bool | true
+generate_for | [InputSet](#inputset) | `**`
+options | [BuilderOptions](#builderoptions) | none
```
Comment (natebosch) [inline docs/build_yaml_format.md]: I don't think this link goes anywhere?

File: docs/build_yaml_format.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+`name`.
+
+To construct a key, you join the package and name with a `:`, so for instance
+the `bar` target in the `foo` package would be referenced like this `foo:bar`.
+
+When referring to targets in the current package, you can omit the package and
+start with a `:` instead, so the above example could be shorted to `:bar`.
+
+Additionally, for any target you can omit the target name, and it will be
+defaulted to the package name, so `foo` would become `foo:foo`.
```
Comment (natebosch) [inline docs/build_yaml_format.md]: We could consider removing some of the magic... this description is not telling the whole story. We perform different normalization when a target is _defined_ from when it is _referenced_. This description is of the reference. In a definition `foo` would become `<current package>:foo`.


Same thing with the builder below. As a definition `foo` becomes `<current package>|foo`, but as a reference it becomes `foo|foo`.

File: build_config/README.md (hunk 1 above)
Comment (jakemac53) [reply build_config/README.md]: gah! ok :(

File: docs/build_yaml_format.md (hunk 2 above)
Comment (jakemac53) [reply docs/build_yaml_format.md]: done

File: docs/build_yaml_format.md (hunk 3 above)
Comment (jakemac53) [reply docs/build_yaml_format.md]: hmm true, I will just remove these paragraphs for now.

**Pull request merged**